src/phoenix/core/model_schema.py

```diff
@@ -0,0 +1,1067 @@
+from random import random
+from typing import (
+    Any,
+    BinaryIO,
+    Callable,
+    Dict,
+    Hashable,
+    Iterable,
+    Iterator,
+    List,
+    Mapping,
+    NamedTuple,
+    Optional,
+    Sequence,
+    Set,
+    Tuple,
+    Type,
+    TypeVar,
+    Union,
+    cast,
+    get_args,
+    overload,
+)
+from uuid import uuid4
+from weakref import ProxyType, proxy
+
+import numpy as np
+import pandas as pd
+from pandas.core.dtypes.common import (
+    is_datetime64_any_dtype,
+    is_datetime64tz_dtype,
+    is_numeric_dtype,
+)
+from typing_extensions import TypeAlias, TypeGuard
+from wrapt import ObjectProxy
+
+
+class DimensionRole(IntEnum):
+    ...
+
+
+@unique
+class SingularDimensionalRole(DimensionRole):
+    UNASSIGNED = auto()
+    PREDICTION_ID = auto()
+    TIME = auto()
+    PREDICTION_LABEL = auto()
+    PREDICTION_SCORE = auto()
+    ACTUAL_LABEL = auto()
+    ACTUAL_SCORE = auto()
+
+
+@unique
+class MultiDimensionalRole(DimensionRole):
+    FEATURE = 1 + len(SingularDimensionalRole)
+    TAG = auto()
+
+
+# global vars for ease of reference
+UNASSIGNED = SingularDimensionalRole.UNASSIGNED
+PREDICTION_ID = SingularDimensionalRole.PREDICTION_ID
+TIME = SingularDimensionalRole.TIME
+PREDICTION_LABEL = SingularDimensionalRole.PREDICTION_LABEL
+PREDICTION_SCORE = SingularDimensionalRole.PREDICTION_SCORE
+ACTUAL_LABEL = SingularDimensionalRole.ACTUAL_LABEL
+ACTUAL_SCORE = SingularDimensionalRole.ACTUAL_SCORE
+FEATURE = MultiDimensionalRole.FEATURE
+TAG = MultiDimensionalRole.TAG
+
+
+def _rand_str() -> str:
+    """Generates a random string, useful for adding a new column to a
+    dataframe, such that it won't conflict with the existing column names.
+    """
+    return hex(int(random() * 1e9))
+
+
+@dataclass(frozen=True)
+class SchemaSpec(ABC):
+    def __post_init__(self) -> None:
+        """Ensure column names are string at run time, which may not be true
+        if the user takes them directly from `pd.dataframe.columns`, which
+        can contain numbers. Phoenix always uses string to refer to columns
+        and dimensions, so this step is intended to eliminate any potential
+        issues caused by any pd.dataframe having numbers as column names.
+        """
+        for f in fields(self):
+            if f.type is str or set(get_args(f.type)) <= {str, type(None)}:
+                # Ensure string if type is `str` or `Optional[str]` (the
+                # latter being `Union[str, None]` under the hood). Use
+                # `__setattr__` because `self` is read-only.
+                object.__setattr__(self, f.name, _ensure_str(getattr(self, f.name)))
+
+
+@dataclass(frozen=True)
+class CompositeDimensionSpec(SchemaSpec, ABC):
+    """A dimension referencing multiple columns (besides its primary column)."""
+
+
+@dataclass(frozen=True)
+class Embedding(CompositeDimensionSpec):
+    vector: str
+    vector_length: Optional[int] = None
+    raw_data: Optional[str] = None
```

Comment:
Maybe we can rename this to just `data` or `text`. Probably more intuitive to a user.